### DISCUSSION TITLE
Unactuated Spring in MuJoCo

### DISCUSSION THREAD

**Discussion by dkhor2003** [google-deepmind/mujoco#2091] [Asking for Help]
### Intro

Hi!

I am a beginner in MuJoCo and Robotics, and I am trying to model the SLIP model in MuJoCo.

### My setup

<mujoco model="SLIP">
    <asset>
        <texture name="texplane" type="2d" builtin="checker" rgb1="1 1 1" rgb2="0 0 0" width="512" height="512"/>
        <material name="MatPlane" texture="texplane" texrepeat="1 1" texuniform="true"/>
    </asset>

    <visual>
        <headlight ambient="0.5 0.5 0.5"/>
    </visual>
    
	<option timestep="0.001" integrator="RK4" gravity="0 0 -9.81">
		<flag contact="enable" energy="enable"/>
	</option>
    
    <worldbody>
        <body name="ground" pos="0 0 0">
            <geom type="plane" material="MatPlane" size="10 10 0.1"/>
        </body>

        <body name="torso" pos="0 0 4">
            <geom type="sphere" size="0.1" rgba="0.3 0.6 1 1"/>
            <joint name="free_body" type="free" pos="0 0 0"/>
            <body name="leg" pos="0 0 -0.1">
                <joint name="hip_x" type="hinge" pos="0 0 0.1" axis="1 0 0" />
    			<joint name="hip_y" type="hinge" pos="0 0 0.1" axis="0 1 0" />
                <geom type="capsule" size="0.05 0.5" fromto="0 0 0 0 0 -0.5" rgba="1 0 0 1"/>
                <geom type="capsule" size="0.05 0.5" fromto="0 0 -0.5 0 0 -0.9" rgba="1 0 0 0.1"/>
                <site name="spring_end_1" pos="0 0 -0.5" size="0.04"/>
                <site name="spring_end_2" pos="0 0 -0.9" size="0.04"/>
            </body>
        </body>
    </worldbody>

    <tendon>
        <spatial name="spring" stiffness="10" damping="1.0" solreflimit="100 0.001">
            <site site="spring_end_1"/>
            <site site="spring_end_2"/>
        </spatial>
    </tendon>

    <actuator>
		<motor ctrllimited="true" ctrlrange="-2.0 2.0" gear="20" joint="hip_x"/>
		<motor ctrllimited="true" ctrlrange="-2.0 2.0" gear="20" joint="hip_y"/>
	</actuator>
    
</mujoco>

### My question

This is my current model xml file. I wanted to use spatial tendon to simulate a spring for the leg of the SLIP model. However, when I drop it from a height, it doesn't look like it is bouncing continuously. Are there any advice on how I can make this work such that the model bounces continuously when dropped from a height and eventually stops? Thanks for any help provided!

### Minimal model and/or code that explain my question

If you encountered the issue in a complex model, please simplify it as much as possible (while still reproducing the issue).

Model:
<details>
<summary>minimal XML</summary>

```XML

<mujoco>
  <worldbody>
    <light pos="0 0 1"/>
    <geom type="sphere" size="1" rgba="1 0 0 1"/>
  </worldbody>
</mujoco>

```
</details>

Code:
```python
import mujoco
import mediapy as media
model = mujoco.MjModel.from_xml_string(xml)
data = mujoco.MjData(model)

with mujoco.Renderer(model) as renderer:
  mujoco.mj_forward(model, data)
  renderer.update_scene(data)

  media.show_image(renderer.render())
```

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment]
1. You don't need a tendon.
2. What you do need is a linear dof in the leg.

Something like 

```xml
...
        <body name="upper leg" pos="0 0 -0.1">
            <joint name="hip_x" type="hinge" pos="0 0 0.1" axis="1 0 0" />
			<joint name="hip_y" type="hinge" pos="0 0 0.1" axis="0 1 0" />
            <geom type="capsule" size="0.05 0.5" fromto="0 0 0 0 0 -0.5" rgba="1 0 0 1"/>
            <body name="lower leg" pos="0 0 -0.5">
              <joint spring type="slide" stiffness="X" damping="Y"/>
              <geom type="capsule" size="0.05 0.5" fromto="0 0 0 0 0 -0.5" rgba="1 0 0 0.1"/>
            </body>
        </body>
...
```

**Reply by dkhor2003** [reply]
@yuvaltassa Thanks for the advice. 

I am trying to make a SLIP model with 8 DOF (6 for rigid body and 2 for hip rotation in x and y direction). If I made the spring as a slide joint in z-axis, wouldn't that make it a 9 DOF model? I am still learning on how to count DOFs, so please correct me if I am wrong.

Thanks.

**Reply by yuvaltassa** [reply]
I suppose one could imagine doing this without a dof actually. You could make the floor contact super soft and springy. This would look like the leg deeply penetrating the floor. But then friction would also be rather weird.

TBH these non-physical approximations are not what MuJoCo was made for. You might be better off writing the equations by hand if you really need the abstraction from the literature
